web/book/src/internals/modules.md

```diff
@@ -0,0 +1,324 @@
+# Modules
+
+> Status: proposal
+
+> This is a technical document. For a "how to use" or a TLDR; skip to the
+> [Example](#Example) section.
+
+Design goals:
+
+1. Allow importing declarations from other files.
+
+2. Have namespaces for things like `std`.
+
+3. Have a hierarchical structure so we can represent files in directories.
+
+4. Have unambiguous module structure within a project.
+
+## Definition
+
+A module is a namespace that contains declarations. A module is itself a
+declaration , which means that it can contain nested child modules.
+
+This means that modules form a
+[tree graph](<https://en.wikipedia.org/wiki/Tree_(graph_theory)>), which we call
+"the module structure".
+
+For the sake of this document, we will express the module structure with
+`module` keyword and a code block encased in curly braces:
+
+```
+module my_playlists {
+    let bangers = ... # a declaration
+
+    module soundtracks {
+        let movie_albums = ... # another declaration
+    }
+}
+```
+
+> The syntax `module name { ...decls... }` is not part of PRQL language, with
+> the objection that it is unnecessary as it only adds more ways of defining
+> modules. If a significant upside of this syntax is found, it may be added in
+> the future.
+
+## Name resolution
+
+Any declarations within a module can be referenced from the outside of the
+module:
+
+```prql no-eval
+# using module structure declared above
+
+let great_tracks = my_playlists.bangers
+
+let movie_scores = my_playlists.soundtracks.movie_albums
+```
+
+Identifiers are resolved relative to current module.
+
+```prql no-eval
+module my_playlists {
+    module soundtracks {
+        let movie_albums = (from albums | filter id == 3)
+    }
+
+    from soundtracks.movie_albums
+}
+from my_playlists.soundtracks.movie_albums
+```
+
+If an identifier cannot be resolved relative to the current module, it tries to
+resolve relative to the parent module. This is repeated, stepping up the module
+hierarchy until a match is found or root of the module structure is reached.
+
+```prql no-eval
+module my_playlists {
+    let decl_1 = ...
+
+    module soundtracks {
+        let decl_2 = ...
+    }
+
+    module upbeat_rock {
+        let decl_3 = ...
+
+        from decl_1 | join soundtracks.decl2 | join decl_3
+    }
+}
+```
+
+## Main var declaration
+
+If the final declaration in a module is a variable declaration that is named
+`main`, then the leading `let main = ` can be omitted and expressed only by the
```

Comment:
Looking at the example — should this instead be "The final statement in a module is implicitly named `main`"? In the example there is no variable declaration named `main`...

Reply:
That may be written poorly, I wanted to say that:
- if the last declaration in a module is a variable declaration,
- and it is named `main`,
- it may omit the leading `let main = `.

It can also be named something else, but it needs to have the `let x =` in this case.

Another way to say it, would be that:

```
Last unnamed pipeline in a module is implicitly named `main` 
```
... but this implies that we can have unnamed pipelines elsewhere in a module.

(ps: I've been using term "declaration" as a cover term for variable, function and type declarations. We can switch that to "statement" if you feel it is more appropriate or future-proof)

Reply:
> * if the last declaration in a module is a variable declaration,
> 
>     * and it is named `main`,
> 
>     * it may omit the leading `let main = `.

I think this is the sticking point... it seems like the conclusion is required for the premise, since it can't be named `main` if it omits the leading `let main =`...
(sorry if I'm misunderstanding — my guess is that this is actually a small point of writing rather than some fundamental disagreement)


> ```
> Last unnamed pipeline in a module is implicitly named `main` 
> ```
> 
> ... but this implies that we can have unnamed pipelines elsewhere in a module.

How about a "An unnamed pipeline as the module's final declaration is implicitly named `main`" ?

Reply:
Oh, I see the problem! In my mind, the last pipeline remains named `main` even if you remove the leading `let main =`. That may be confusing, yes.

> How about a "An unnamed pipeline as the module's final declaration is implicitly named main" ?

That is still implying there are unnamed pipelines.

Last variable declaration in a module can omit the leading `let main =` and acquire an implicit name `main`.

(yes this appears to be a minor point in writing)

Reply:
> Last variable declaration in a module can omit the leading `let main =` and acquire an implicit name `main`.

Perfect

(nitty: I'd use "final" rather than "last" as it's less ambiguous, though here it's fairly clear given the "in a module" modifier)